PENG-2369 implement the license manager agent snap

matheushent opened the pull request:
#### What
This PR implement a snap for installing and running the license manager agent. As well as, it modifies the workflows so quality assurance is run against the snap sub project. Also, a new job is added in the *Publish on Tag* workflow for publishing the snap to the *latest/candidate* and *latest/edge* channels.

#### Why
This PR facilitates the installation and deployment of the agent

`Task`: https://sharing.clickup.com/t/h/c/18022949/PENG-2369/RLHZFG26P6LMK8P

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

matheushent committed:
PENG-2369 implement the license manager agent snap, as well as modify the workflows for interacting with it

dusktreader commented on lm-agent-snap/Makefile:
```diff
@@ -0,0 +1,33 @@
+qa: lint type ## Run all quality assurance checks
+
+.PHONY: format
+type: ## Run the formatter
```
```suggestion
format: ## Run the formatter
```

matheushent replied:
fixed in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/README.md:
```diff
@@ -0,0 +1,76 @@
+
+## Basic Usage
+
+This snap requires a few configuration values to be set before it can be used. These values are:
```
I think we should change the wording here since most of these settings are not required to be changed. It might be good to break up the settings into two groups. The first could be the required settings (`oidc-client-id` and `oidc-client-secret`) and the second list could be the optional settings. What do you think?

matheushent replied:
that's a good idea, thank you

matheushent replied:
done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/README.md:
```diff
@@ -0,0 +1,76 @@
+
+# Installation instructions
+
+For installing from the remote store, run:
```
What do you think of having this say "Snap store" instead and make it a link?

matheushent replied:
done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/README.md:
```diff
@@ -0,0 +1,76 @@
+
+- single-user-submitter: The system user name that the agent will use to submit jobs to the SLURM cluster on behalf of in case the *single-user-submitter* is not set. This is optional and defaults to *ubuntu*.
+
+- write-submission-files: A boolean value (true, false) that indicates whether the agent should write submission files to disk. This is optional and defaults to false.
```
This needs to be re-written for the license manager config settings.

matheushent replied:
damn, I forgot to copy from the snapcraft.yaml file

matheushent replied:
fixed in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/README.md:
```diff
@@ -0,0 +1,76 @@
+
+# Development
+
+For development purposes, you can build the `license-manager-agent` part prior of packing the snap. To do that, run:
```
```suggestion
For development purposes, you can build the `license-manager-agent` part prior to packing the snap. To do that, run:
```

matheushent replied:
done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/hooks/bin/configure:
```diff
@@ -0,0 +1,84 @@
+    return snapctl_out
+
+
+def run_bash(bash_string: str) -> str:
```
It would be good to move this to the top of the file since it is used by functions that are defined before it.

matheushent replied:
modified in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

dusktreader commented on lm-agent-snap/hooks/bin/configure:
```diff
@@ -0,0 +1,84 @@
+
+    Return python None if snapctl returns the empty string.
+    """
+    snapctl_out: Union[str, None]
```
Why use the old style type definintion here instead of `str | None`?

matheushent replied:
I don't know why but the type checker failed sometimes for the vantage agent. Here I just copied tbh.

matheushent replied:
I also remember having errors after the snap got installed due to python versions not accepting this. For example, installing the snap on ubuntu 20.04 would fail

dusktreader replied:
I guess this must be due to the snap using the system python. For 20.04, it's Python 3.8. So, this would explain why there are problems. Thanks for the explanation!

dusktreader commented on lm-agent-snap/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,80 @@
+
+  - reservation-identifier: Reservation name used when doing the reconciliation. Defaults to *license-manager-reservation*.
+
+  - lm_user: System user name for creating reservations on behalf of. Defaults to *license-manager*.
```
```suggestion
  - lm_user: System user name used when creating reservations. Defaults to *license-manager*.
```

matheushent replied:
modified in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

matheushent committed:
PENG-2369 modifications for PR review

This commit include modifications for improving texting in the README.md and slight improves for better readability

julianaklulo commented on lm-agent/lm_agent/config.py:
```diff
@@ -10,7 +10,7 @@
 
 
-DEFAULT_DOTENV_PATH = Path("/etc/default/license-manager-agent")
+DEFAULT_DOTENV_PATH = Path("/var/snap/license-manager-agent/common/.env")
```
Is this change really necessary?
In Scania's environment the `.env` file is loaded from the default path, the charm is configured to populate the env vars in there.

matheushent replied:
this is needed according to our intention in making the agent installable by snap instead of configuring the virtual environment, installing the python package, etc...

dusktreader replied:
Damn...Juliana is right, this will cause issues for our deployments of lm-agent on non-Ubuntu machines. We have some deployments on Rocky 9 and those environments do not have snap installed (and we will not be able to install it, either).

Instead of having one path we look in for the dotenv file, perhaps we need to have a set of places that we look in order until we find a dotenv file. What do you both think of this?

matheushent replied:
we discussed in private that we would move with this implementation. Can we resolve @dusktreader @julianaklulo ?

dusktreader replied:
As per our prior discussion, the charm-license-manager-agent is already configuring the systemd service to use an `EnvironmentFile`. We verified that agent when deployed with the charm is able to load its config successfully without needing to set the DEFAULT_DOTENV_PATH to `/etc/default/license-manager-agent`. Resolving this comment since we are moving forward as planned and tested.

matheushent committed:
PENG-2369 add apps to the snap for starting, stopping and restarting the agent daemon
